collector/collector.go

```diff
@@ -0,0 +1,607 @@
+					}
+
+					metric := plugin.Metric{
+						Timestamp: time.Now(),
+						Namespace: ns,
+						Data:      c.containers[rid].Stats.Cgroups.CpuStats.CpuUsage.PerCpu[cpuID],
+						Tags:      mt.Tags,
+						Config:    mt.Config,
+						Version:   PLUGIN_VERSION,
+					}
+					metrics = append(metrics, metric)
+				}
+
+			case "hugetlb_stats":
+				if metricName[0] == "*" {
+					for size, hugetlb := range c.containers[rid].Stats.Cgroups.HugetlbStats {
+						rns := make([]plugin.NamespaceElement, len(ns))
+						copy(rns, ns)
+
+						rns[indexOfDynamicElement+lengthOfNsPrefix].Value = size
+						metric := plugin.Metric{
+							Timestamp: time.Now(),
+							Namespace: rns,
+							Data:      utils.GetValueByNamespace(hugetlb, []string{mt.Namespace.Strings()[len(ns)-1]}),
+							Tags:      mt.Tags,
+							Config:    mt.Config,
+							Version:   PLUGIN_VERSION,
+						}
+
+						metrics = append(metrics, metric)
+					}
+				} else {
+					metric := plugin.Metric{
+						Timestamp: time.Now(),
+						Namespace: ns,
+						Data:      utils.GetValueByNamespace(c.containers[rid].Stats.Cgroups.HugetlbStats, mt.Namespace.Strings()[len(ns)-2:]),
+						Tags:      mt.Tags,
+						Config:    mt.Config,
+						Version:   PLUGIN_VERSION,
+					}
+					metrics = append(metrics, metric)
+				}
+
+			case "io_merged_recursive":
+				numOfDevices := len(c.containers[rid].Stats.Cgroups.BlkioStats.IoMergedRecursive) - 1
+				if metricName[0] == "*" {
+					for deviceID, val := range c.containers[rid].Stats.Cgroups.BlkioStats.IoMergedRecursive {
+						rns := make([]plugin.NamespaceElement, len(ns))
+						copy(rns, ns)
+
+						rns[indexOfDynamicElement+lengthOfNsPrefix].Value = strconv.Itoa(deviceID)
+						metric := plugin.Metric{
+							Timestamp: time.Now(),
+							Namespace: rns,
+							Data:      val,
+							Tags:      mt.Tags,
+							Config:    mt.Config,
+							Version:   PLUGIN_VERSION,
+						}
+						metrics = append(metrics, metric)
+					}
+				} else {
+					deviceID, err := strconv.Atoi(metricName[0])
+					if err != nil {
+						return nil, fmt.Errorf("In metric %s the given cpu id is invalid, err=%v", strings.Join(mt.Namespace.Strings(), "/"), err)
+					}
+					if deviceID > numOfDevices || deviceID < 0 {
+						return nil, fmt.Errorf("In metric %s the given cpu id is invalid, expected value in range 0-%d", strings.Join(mt.Namespace.Strings(), "/"), numOfDevices)
+					}
+
+					metric := plugin.Metric{
+						Timestamp: time.Now(),
+						Namespace: ns,
+						Data:      c.containers[rid].Stats.Cgroups.BlkioStats.IoMergedRecursive[deviceID],
+						Tags:      mt.Tags,
+						Config:    mt.Config,
+						Version:   PLUGIN_VERSION,
+					}
+					metrics = append(metrics, metric)
+				}
+			} // the end of switch statsType
+		} // the end of range over ids
+	}
+
+	if len(metrics) == 0 {
```

Comment:
I don't think so. It tests if metrics collection process actually collected anything